### PR TITLE
Update mutect2 pipeline

### PR DESCRIPTION
**Pull request by mgeaghan-garvan**
- Added TOML configuration files for submission via CPG hail batch
- Fixed minor bugs in R code when testing for NAs
- Updated CPG hail batch submission scripts
- Updated whitelist dockerfile to include R scripts within the container
- Updated WDL code for loading annovar database
- Added a single job version of the Mutect2 pipeline (optimised for DNAnexus runs)

### CODE REVIEW

File: docker/annovar/Dockerfile
Comment (MattWellie) [inline docker/annovar/Dockerfile:1]: could we move this to `https://github.com/populationgenomics/images` to make use of the CI to build

File: docker/annovar/README
Comment (MattWellie) [inline docker/annovar/README:5]: ah... maybe not

File: scripts/cpg_cromwell/m2.sj.submit.py
Comment (MattWellie) [inline scripts/cpg_cromwell/m2.sj.submit.py]: a good chunk of this should be swapped out for a call of `cpg_utils.hail_batch.get_batch`

File: scripts/cpg_cromwell/m2.submit.py
Comment (MattWellie) [inline scripts/cpg_cromwell/m2.submit.py]: same here, cpg_utils' get_batch wrapper makes this much cleaner

File: scripts/cpg_cromwell/pon.submit.py
Comment (MattWellie) [inline scripts/cpg_cromwell/pon.submit.py]: image

**Review by MattWellie — COMMENTED** [review]
I've only skimmed the WDL/R stuff. Main useful change is the migration to the cpg_utils Batch wrapper, and I'm not sure if moving/duplicating the dockerfiles into the CPG images repo is practical

File: docker/annovar/README
Comment (mgeaghan) [reply docker/annovar/README:5]: Yeah, so Annovar is a bit weird in that to download it you're supposed to go through a (free) registration process, and you get emailed a link to the download. Very archaic! It does look like the links stay active indefinitely though, so I could update the Dockerfile to download the package itself, but it could break at any time.

File: scripts/cpg_cromwell/m2.sj.submit.py
Comment (mgeaghan) [reply scripts/cpg_cromwell/m2.sj.submit.py]: I'll update this to use the new get_batch wrapper, but I still need the dataset and access level information to pass to run_cromwell_workflow. So it cleans it up a little but not by a whole lot.

**Comment by mgeaghan** [comment]
@MattWellie I've updated the submission scripts to use the cpg_utils get_batch function as you suggested. However, as per my comment above, it doesn't clean up too much as I still need much of the config information for submitting the jobs to cromwell.
I've also added a data export script that moves the results of the pipeline to tmp. This is so we can export the data to the private buckets of Owen's lab. Is this acceptable wrt the CPG data policies? If not, we should set up a meeting with Owen to work out a suitable data storage strategy.

File: docker/annovar/README
Comment (MattWellie) [reply docker/annovar/README:5]: Yeah, maybe that's not worth doing then. We could use `skopeo` to copy the built image directly into our artifactory, so long as there are no licensing implications

**Comment by MattWellie** [comment]
> I've also added a data export script that moves the results of the pipeline to tmp. This is so we can export the data to the private buckets of Owen's lab. Is this acceptable wrt the CPG data policies? If not, we should set up a meeting with Owen to work out a suitable data storage strategy.

This shouldn't be required... the cromwell processes should generate files in tmp, then copies specified files back into permanent storage. That would mean they already sit in tmp upon completion, so no copy required. 

In terms of exporting to Owen's Lab's buckets, tmp has the same read restrictions as the main bucket, so this won't make them more accessible. You can write to the `main-analysis` bucket instead, which will grant ability to read data direct from the bucket. Otherwise there's a `main-release` bucket for this purpose, which would be followed by granting permission to specific users to copy data away using limited-duration access. I've no idea who/what is required to deal with the access approval for copying the data away, but this script may need some work. Maybe we can schedule a chat before deciding how to alter this?

**Comment by mgeaghan** [comment]
@MattWellie Thanks for that explanation. I thought tmp had read access, my bad. Happy then to move the outputs to analysis so we can copy out results, but I think then I will make an additional script to delete from analysis once they've been copied so we don't double up on storage costs.

**Comment by MattWellie** [comment]
I used to think that too 😓  makes it a little frustrating when cromwell log files are copied over to `-tmp` and still aren't always visible.

For this purpose the [release](https://github.com/populationgenomics/team-docs/tree/main/storage_policies#release-gscpg-dataset-release) bucket is the most appropriate solution. I'll tag @illusional who will have more experience with providing external access using this method.

**Comment by mgeaghan** [comment]
Ok, that sounds good, In that case we'll need to get a release bucket set up for the kccg-genomics-med project, as one doesn't exist yet.

File: docker/annovar/README
Comment (illusional) [reply docker/annovar/README:5]: Could you put the download on a bucket, and then copy it in? I guess that requires creds thought which isn't great.

**Comment by mgeaghan** [comment]
@MattWellie I updated the script to use the release bucket. Currently there's no neat programmatic way to do this, as the release buckets aren't present in the TOML configs, so I'm using a string replace to exchange '-main' for '-release' for now.
I also added a deletion script for the cases where we export the pipeline outputs (e.g. exporting to buckets for Owen's group) and want to remove the outputs from the CPG buckets to reduce double charging. It's essentially the same script just with a gsutil rm command instead.

**Comment by MattWellie** [comment]
> Currently there's no neat programmatic way to do this, as the release buckets aren't present in the TOML configs

Ugh, I hadn't realised that. Best guess would be because this is an optional dataset (when projects/accounts/buckets are created, creating a corresponding release bucket is a toggle that defaults to False). That's out of scope for this change, but will look at that

File: scripts/cpg_cromwell/export_analysis_outs.py
Comment (MattWellie) [inline scripts/cpg_cromwell/export_analysis_outs.py:51]: @illusional will this work? From memory writing data to release should be fine, for reading the billing project will have to be set specifically in the command

File: scripts/cpg_cromwell/export_analysis_outs.py
Comment (MattWellie) [inline scripts/cpg_cromwell/export_analysis_outs.py:29]:
```suggestion
b.run(wait=False)

```

File: scripts/cpg_cromwell/remove_analysis_outs.py
Comment (MattWellie) [inline scripts/cpg_cromwell/remove_analysis_outs.py:48]: @illusional likewise - would the billing account need to be set specifically here to delete from a requestor-pays bucket?

Either way this operation would require running using `-access-level full` through analysis-runner

**Review by MattWellie — APPROVED** [review]
Couple of Qs but looks good

File: scripts/cpg_cromwell/remove_analysis_outs.py
Comment (illusional) [reply scripts/cpg_cromwell/remove_analysis_outs.py:48]: ~~Yes definitely, can we make this process happen across multiple lines to improve readability?~~

I think deleting here is actually fine.

File: scripts/cpg_cromwell/remove_analysis_outs.py
Comment (illusional) [inline scripts/cpg_cromwell/remove_analysis_outs.py:1]: This script is getting close to the border of being too permissive to delete and move files to release because it relies on user input here. As it allows moving to and deleting from release and this repository and datasets this belongs to is within your control, I'm happy to leave it, but see the comment below.

File: scripts/cpg_cromwell/remove_analysis_outs.py
Comment (illusional) [inline scripts/cpg_cromwell/remove_analysis_outs.py:38]: Security wise, this is prone to injection attacks, say the output prefix:

```
--output-dir '; gsutil cp gs://cpg-<dataset>-main/crams/ gs://other-bucket'
```

You'll delete everything in the bucket, with no recourse to restore because `rm -r` deletes all versions.

**Review by illusional — CHANGES_REQUESTED** [review]

File: scripts/cpg_cromwell/export_analysis_outs.py
Comment (illusional) [reply scripts/cpg_cromwell/export_analysis_outs.py:51]: Writing is a Class A operation, so pretty sure it requires a billing account:

image

File: scripts/cpg_cromwell/remove_analysis_outs.py
Comment (mgeaghan) [reply scripts/cpg_cromwell/remove_analysis_outs.py:38]: Very good point, apologies for the oversight. I'll get onto fixing that this afternoon.

Comment (mgeaghan) [reply scripts/cpg_cromwell/remove_analysis_outs.py:38]: @illusional I've made a few changes to the code to restrict all the final paths to the regex `^gs://[A-Za-z0-9\-_\/\.]+$`. That should ensure that no commands can be injected into the paths. I've also added a list of valid projects so the buckets being worked on can only be in that list. Right now that just includes kccg-genomics-med. Any new projects will need a code review to update these scripts.

Comment (illusional) [reply scripts/cpg_cromwell/remove_analysis_outs.py:38]: Amazing, thanks @mgeaghan!

**Review by illusional — APPROVED** [review]
Approving to dismiss my request changes, but leaving to @MattWellie to actually review.

**Pull request merged**
